moby/moby
# Add parallel gzip support controlled by a daemon flag

Signed-off-by: Jon Botelho <jbotelho2@bloomberg.net>



**- What I did**
Improved image push performance for large images on fast networks, where compression is the bottleneck.

**- How I did it**
Added a `--compression-threads` flag on the daemon, which gzip compression [pgzip](https://github.com/klauspost/pgzip) for multi-threaded compression of the push of `--compress-threads` is set greater than 1 (eg. `--compression-threads=2`). When you set `--compression-threads=1`, it will fall back to the original behavior of using `compress/gzip`. This is pretty much a continuation of the work and discussion in #17674.

I'm new to this project, so I'm not sure the way I've implemented this flag is ideal. I could also use some guidance on what kind of test coverage this change would need.

**- How to verify it**
Start up the daemon in debug mode with `--compression-threads=2` (or above), and push an image. Observe that it pushes correctly, and observe that the log mentions its using pgzip.

**- Description for the changelog**

TBD

**- A picture of a cute animal (not mandatory but encouraged)**
Containerized dog:
![container_dog](https://cloud.githubusercontent.com/assets/13650269/22768628/58ee2584-ee51-11e6-96ed-5d99e8e486b5.jpg)
